R/tdata.R

```diff
@@ -0,0 +1,99 @@
+#' Create a new `tdata` object which contains
+#'
+#' - `data` a `reactive` list of data.frames with attributes
+#' i) `code` (`reactive`) containing code used to generate the data
+#' and ii) join_keys (`JoinKeys`) containing the relationships between
+#' the data
+#'
+#' @param data A `named list` of `data.frames` which optionally can be `reactive`.
+#'   Inside this object all `data.frames` will be made `reactive`.
+#' @param code A `character` (or `reactive` which evaluates to a `character`) containing
+#'   the code used to generate the data. This should be `reactive`if the code is changing
+#'   during a reactive context (e.g. if filtering changes the code). Inside this
+#'   object `code` will be made reactive
+#' @param join_keys A `teal.data::JoinKeys` object containing relationships between the
+#'   datasets.
+#'
+#' @return A `tdata` object
+#' @examples
+#'
+#' data <- new_tdata(
+#'   data = list(iris = iris, mtcars = reactive(mtcars)),
+#'   code = "iris <- iris
+#'     mtcars = mtcars"
+#' )
+#'
+#' # Extract a data.frame
+#' isolate(data[["iris"]]())
+#'
+#' @export
+new_tdata <- function(data, code = "", join_keys = NULL) {
+  checkmate::assert_list(
+    data,
+    types = c("data.frame", "reactive"), any.missing = FALSE, names = "unique"
+  )
```

Comment:
an additional validation to check the if the content of the reactive is a `data.frame` would be good to avoid such entries:
`data <- list(d1 = reactive(1), d2 = reactive("A"))`

Reply:
yup I'll add that (though think it makes sense to do it whilst looping through contents a couple of lines down

Reply:
so I've had to remove the validation here as in the MAE/helios case data[[x]] does not have to be a data.frame - oops

Reply:
@Polkas @mhallal1 - not sure what to do here

Reply:
actually I am :)